### DISCUSSION TITLE
Windows 10 - Animation memory leak

### DISCUSSION THREAD

**Discussion by Xavman42** [DigiScore/neoscore#50] [General]
I seem to be experiencing a memory leak when running the sample animation script in the documentation [here](https://neoscore.org/core_concepts/interactivity.html). My RAM fills up at a rate of ~300 MB per second at 10 fps, and that rate becomes faster as I increase fps. Is there any memory management I can do to fix this?

I am using Windows 10 and Python 3.10.5.

Thank you for your time.

````
from neoscore.common import *
import math

neoscore.setup()
center = Mm(50)
text = Text((center, Mm(50)), None, "moving text")

def refresh_func(time: float):
    text.x = center + Mm(math.sin(time) * 20)

neoscore.show(refresh_func, 10)

**Comment by ajyoon** [comment]
Thanks for the report! That definitely sounds like a memory leak. I don't get that on Linux, so I would guess it might be a bad assumption in our Qt layer or our homemade text path caching. I know Qt handles fonts pretty differently on Windows.

I can dig into this on a VM later, but in the meantime if you'd be able to run a memory profiler over that and share the results it would help us find the bug quicker. Otherwise I can try to reproduce on my end later.

-------- Original Message --------

**Comment by Xavman42** [comment]
Sure thing! Here's my implementation of the memory_profiler:

```
from neoscore.common import *
import math
from memory_profiler import profile

neoscore.setup()
center = Mm(50)
text = Text((center, Mm(50)), None, "moving text")

@profile
def refresh_func(time: float):
    text.x = center + Mm(math.sin(time) * 20)

neoscore.show(refresh_func, 10)
```

Here are the first two outputs:
```
Line #    Mem usage    Increment  Occurrences   Line Contents
=============================================================
     9    114.8 MiB    114.8 MiB           1   @profile
    10                                         def refresh_func(time: float):
    11    114.8 MiB      0.0 MiB           1       text.x = center + Mm(math.sin(time) * 20)


Line #    Mem usage    Increment  Occurrences   Line Contents
=============================================================
     9    173.7 MiB    173.7 MiB           1   @profile
    10                                         def refresh_func(time: float):
    11    173.7 MiB      0.0 MiB           1       text.x = center + Mm(math.sin(time) * 20)
```

And here are the last two outputs:
```

Line #    Mem usage    Increment  Occurrences   Line Contents
=============================================================
     9   2057.5 MiB   2057.5 MiB           1   @profile
    10                                         def refresh_func(time: float):
    11   2057.5 MiB      0.0 MiB           1       text.x = center + Mm(math.sin(time) * 20)


Line #    Mem usage    Increment  Occurrences   Line Contents
=============================================================
     9   2058.9 MiB   2058.9 MiB           1   @profile
    10                                         def refresh_func(time: float):
    11   2058.9 MiB      0.0 MiB           1       text.x = center + Mm(math.sin(time) * 20)

Process returned 0 (0x0)        execution time : 14.391 s
```

I'm not sure how informative this is, but I hope it helps somehow!

**Comment by ajyoon** [comment]
Ok, on my VM I can only reproduce a much smaller memory leak of about 0.5MB/s, and I can't reproduce any leak on my Linux host machine. The leak you're experiencing must be of a different nature than what I could reproduce since it's so much larger. I'm guessing Qt-related issues. Could you please run [this modified example](https://pastebin.com/vtdnUN5R) and upload the profile printed at the end?

**Comment by Xavman42** [comment]
Here you go! I'm not using Python best-practices and using a VM, but I wouldn't expect that to contribute to this particular issue.
[Profile.txt](https://github.com/DigiScore/neoscore/files/8987333/Profile.txt)

**Reply by ajyoon** [reply]
Hmm, that profile is only showing about 24mb allocated. Could you confirm that when you run that you're still seeing memory explode on task manager or whatever?

**Reply by Xavman42** [reply]
Yeah, when I monitor task manager, Python memory goes up by ~200MB per second. (I imagine task manager refreshes every second) During this test, Python eats up a bit more than 1,700MB according to task manager. I'll run that other script in a moment.

**Comment by ajyoon** [comment]
Also, could you please run [this other profiler-instrumented script?](https://pastebin.com/gUAFYfE3). It requires installing pympler first with `pip install pympler`

**Reply by Xavman42** [reply]
Here you go!

```
                          types |   # objects |   total size
=============================== | =========== | ============
                           dict |       27431 |      9.19 MB
                            str |       58336 |      7.77 MB
                           code |       17345 |      2.97 MB
                           type |        2688 |      2.38 MB
                           list |       13417 |      1.19 MB
                          tuple |       20592 |      1.19 MB
                            set |        1840 |    699.88 KB
                sip.wrappertype |         662 |    691.98 KB
                          float |       27626 |    647.48 KB
                   sip.enumtype |         560 |    586.25 KB
             wrapper_descriptor |        5997 |    421.66 KB
                        weakref |        4889 |    343.76 KB
  neoscore.core.units.StaffUnit |        5360 |    335.00 KB
                    abc.ABCMeta |         277 |    284.04 KB
              getset_descriptor |        3487 |    217.94 KB

Process returned 0 (0x0)        execution time : 17.548 s
```

**Comment by ajyoon** [comment]
Ok it seems very likely to be on the Qt side. It might have to do with different font handling on some windows environments. Could you try running this simpler example which uses simple paths instead of text and let me know if you still see the leak?

```python
import math

from neoscore.core import neoscore
from neoscore.core.path import Path
from neoscore.core.point import ORIGIN
from neoscore.core.units import ZERO, Mm

neoscore.setup()

p1 = Path.rect(ORIGIN, None, Mm(2), Mm(1))
p2 = Path.rect(ORIGIN, None, Mm(1), Mm(3))

center = ZERO


def refresh_func(time):
    p1.x = Mm(math.sin((time / 2)) * 10)
    p2.x = Mm(math.sin((time / 2) + 1) * 12)


if __name__ == "__main__":
    neoscore.show(refresh_func)

```

**Reply by Xavman42** [reply]
No luck yet, the leak is still very present when running this script.

**Comment by ajyoon** [comment]
Sorry for the slow progress here, this is a tough one. I just tried reproducing on an actual Windows 10 laptop and couldn't see the leak there either. What version of Windows 10 are you on? Are you using the Windows Subsystem for Linux by any chance? I don't think it's likely I'll be able to crack this without reproducing on my end.

**Comment by Xavman42** [comment]
No worries! If it's just a me issue, then maybe I'll just dual-boot Win 10 and Ubuntu and call it a day.
Here are my Windows details:

Device name	LAPTOP-482T2PL
Processor	Intel(R) Core(TM) i5-8300H CPU @ 2.30GHz   2.30 GHz
Installed RAM	8.00 GB (7.86 GB usable)
Device ID	E2BEE0DB-1724-43BA-B7C8-67A68C9BBB2C
Product ID	00325-96338-81589-AAOEM
System type	64-bit operating system, x64-based processor
Pen and touch	No pen or touch input is available for this display

Edition	Windows 10 Home
Version	21H2
Installed on	‎12/‎27/‎2021
OS build	19044.1766
Experience	Windows Feature Experience Pack 120.2212.4180.0

**Reply by ajyoon** [reply]
Thanks, I'll see if I can find a VM image which reproduces this better. In the meantime if you feel up for another experiment, I wonder if running on WSL helps? Just a random guess.

**Reply by Xavman42** [reply]
I've run into a couple walls getting QT to work on WSL, but I think it should be possible. I'll let you know if I ever get it working.
